crates/ruff_graph/src/collector.rs

```diff
@@ -0,0 +1,120 @@
+use red_knot_python_semantic::ModuleName;
+use ruff_python_ast::visitor::source_order::{walk_body, walk_expr, walk_stmt, SourceOrderVisitor};
+use ruff_python_ast::{self as ast, Expr, ModModule, Stmt};
+
+/// Collect all imports for a given Python file.
+#[derive(Default, Debug)]
+pub(crate) struct Collector<'a> {
+    /// The path to the current module.
+    module_path: Option<&'a [String]>,
+    /// Whether to detect imports from string literals.
+    string_imports: bool,
+    /// The collected imports from the Python AST.
+    imports: Vec<CollectedImport>,
+}
+
+impl<'a> Collector<'a> {
+    pub(crate) fn new(module_path: Option<&'a [String]>, string_imports: bool) -> Self {
+        Self {
+            module_path,
+            string_imports,
+            imports: Vec::new(),
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn collect(mut self, module: &ModModule) -> Vec<CollectedImport> {
+        walk_body(&mut self, &module.body);
+        self.imports
+    }
+}
+
+impl<'ast> SourceOrderVisitor<'ast> for Collector<'_> {
+    fn visit_stmt(&mut self, stmt: &'ast Stmt) {
+        match stmt {
+            Stmt::ImportFrom(ast::StmtImportFrom {
+                names,
+                module,
+                level,
+                range: _,
+            }) => {
+                let module = module.as_deref();
+                let level = *level;
+                for alias in names {
+                    if level == 0 {
+                        let mut components = vec![];
+
+                        // Add the module path.
+                        if let Some(module) = module {
+                            components.extend(module.split('.'));
+                        }
+
+                        // Add the alias name.
+                        components.push(alias.name.as_str());
+
+                        if let Some(module_name) = ModuleName::from_components(components) {
+                            self.imports.push(CollectedImport::ImportFrom(module_name));
+                        }
+                    } else if let Some(module_path) = self.module_path {
+                        let mut components = vec![];
+
+                        // Start with the containing module.
+                        components.extend(module_path.iter().map(String::as_str));
+
+                        // Remove segments based on the number of dots.
+                        for _ in 0..level {
+                            if components.is_empty() {
+                                return;
+                            }
+                            components.pop();
+                        }
```

Comment:
If I'm reading `to_module_path` correctly, I think for an `__init__.py` module, `self.module_path` will be e.g. `['foo', 'bar', '__init__']` -- that is, it will include `__init__` as a final component?

That's necessary for this logic to be correct, since the meaning of a relative import is one level different in `foo/bar/__init__.py` vs in `foo/bar.py`.

Reply:
Correct!